### ISSUE TITLE
GSoC proposal: MySQL query parallelization

### ISSUE THREAD

**Issue by shlomi-noach** [github/gsoc#94]
#### Before you start

Please read the student manual at http://write.flossmanuals.net/gsocstudentguide/what-is-google-summer-of-code/ and make sure you understand how the program works, how proposals are submitted and all the milestones involved.
#### Proposal

Summary: a proxy to MySQL servers that, given a complex `SELECT` query, breaks it into multiple disjoint queries and issues them in parallel on multiple backends.

Expected Outcome: A 100% correct result (identical, up to sorting, to the result set achieved by the single original query) obtained by evenly distributing query parts on backend pool. A limited set of complex, common production queries is expected to pass the test.
Hopefully this also results with reduced query time for complex queries

Skills: golang, SQL. Will use parts of [Vitess](https://github.com/youtube/vitess)

Mentors: @shlomi-noach

**Comment by hTrap** [comment]
Hey @shlomi-noach, this project sounds really interesting, I would like to work on it. Can you give me some pointers on how I might get started?

**Comment by freddiev4** [comment]
Hey @shlomi-noach! This project sounds super interesting! Over the past few months I've been working with SQL and have been looking for an excuse to brush up on my Golang skills. I don't know much about parallelization but am very interested to learn more via this project. Could you let me know how I could get started or how I could best prepare myself to work on this? 

I found this via GSoC but unfortunately I won't be officially participating this year; I'd absolutely love to get involved though!

**Comment by shlomi-noach** [comment]
Student application is to begin on March 14, if I'm not mistaken. Until such time anyone is a free citizen and can poke around as much as they like; if you get started it's on your own.

**Comment by shlomi-noach** [comment]
As background people can poke into [Vitess](https://github.com/youtube/vitess), and in particular into the [SQL parser](https://github.com/youtube/vitess/tree/master/go/vt/sqlparser).
Also noteworthy is [go-mysql](https://github.com/siddontang/go-mysql#server) where you can find a MySQL server mockup implementation.

**Comment by dinever** [comment]
Hi @shlomi-noach, if I am not mistaken, this will be a stand-alone project. So I can start to create a repository and work on it, right?

**Comment by shlomi-noach** [comment]
Elaborating on this project:

This is a one developer project. Anyone may create a repository and start working on this. Only GSoC applicants will be communicated with, though. GSoC applicants can begin drafting their proposals at their leisure.

More technical description:
- It is a well known fact that MySQL only utilizes a single core to a given query.
- A common setup is to have multiple MySQL backends, in the form of replicas
- Some queries can be split into disjoint queries, and can be executed in parallel. Examples to such queries are:
  - `SELECT * FROM users WHERE user_id IN (2, 3, 5, 8, 2, 13)`
  - `SELECT * FROM users WHERE registered=0 OR validated=0`
  - `SELECT * FROM users JOIN projects USING user_id WHERE users.last_seen >= NOW() - INTERVAL 1 DAY OR projects.last_updated >= NOW() - INTERVAL 1 DAY OR (projects.is_open AND users.is_contributor)`
  - `SELECT user_id FROM users WHERE is_contributor UNION ALL SELECT user_id FROM projects WHERE is_open`

We would like to be able to split queries into a pre-defined number of chunks. Say we have 3 backend MySQLs to work with, or just have 3 spare cores. Can we split a query into 3 disjoint queries? Can we only split it into 2 disjoint queries? Can we only do 1-or-n?

Prove that the disjoint queries are indeed disjoint. The union of the result sets of all disjoint queries must equal the result set of the original query, up to sorting.

Some query syntax cannot be supported. What syntax can you support? What cannot be supported? Identify those cases you are able to support and only break those queries; the rest, just pass through as they came.

At the end of the day, a client wishes to connect to what _it thinks_ is a MySQL server, issue a query, and get a valid and correct result set. The client need not know the query has been split.
We _hope_ to see either better response time or more throughput. We are _content_ and _happy_ with a correct solution.

We skip authentication considerations in this project. We skip questions of high availability (or availability at all).

**Comment by shlomi-noach** [comment]
I am afraid I am no longer available to mentor for this project for this GSoC session. My sincere apologies to the potential students.

**Comment by h4ck3rk3y** [comment]
I was interested in working on this as a side project and not as a GSoC student as my summer wasn't free enough but I wasn't sure if I should go ahead with the idea in case someone else is working on it but now I guess there won't be any conflict as this isn't open for mentorship.

Sad to hear that this won't be a GSoC project though. Was really looking forward to see this implemented, this was by far the most interesting one on the list.
